integration_create_delete_test.go

```diff
@@ -0,0 +1,206 @@
+// to delete that snapshot. Since testing these functions requires a lot of setup in an AWS account, it's faster and
+// easier to do a single integration test rather than a number of smaller unit tests that each do a lot of setup and
+// teardown.
+func TestCreateAndDelete(t *testing.T) {
+	t.Parallel()
+
+	logger := log.New(os.Stdout, "TestCreateAndDelete ", log.LstdFlags)
+
+	session := session.New(&aws.Config{Region: aws.String(AWS_REGION_FOR_TESTING)})
+	svc := ec2.New(session)
+
+	instance, uniqueId := launchInstance(svc, logger, t)
+	defer terminateInstance(instance, svc, logger, t)
+	waitForInstanceToStart(instance, svc, logger, t)
+
+	ui := &cli.BasicUi{
+		Reader:      os.Stdin,
+		Writer:      os.Stdout,
+		ErrorWriter: os.Stderr,
+	}
+
+	snapshotId := takeSnapshot(instance, uniqueId, svc, ui, logger, t)
+	waitForSnapshotToBeAvailable(snapshotId, svc, logger, t)
+	verifySnapshotWorks(snapshotId, svc, logger, t)
+
+	deleteSnapshotForInstance(instance, ui, logger, t)
+	waitForSnapshotToBeDeleted(snapshotId, svc, logger, t)
+	verifySnapshotIsDeleted(snapshotId, svc, logger, t)
+}
+
+func launchInstance(svc *ec2.EC2, logger *log.Logger, t *testing.T) (*ec2.Instance, string) {
+	uniqueId := UniqueId()
+	userData := fmt.Sprint(USER_DATA_TEMPLATE, uniqueId, TEST_FILE_PATH)
+
+	logger.Printf("Launching EC2 instance in region %s. Its User Data will create a file %s with contents %s.", AWS_REGION_FOR_TESTING, TEST_FILE_PATH, uniqueId)
+
+	runResult, err := svc.RunInstances(&ec2.RunInstancesInput{
+		ImageId:      aws.String(AMAZON_LINUX_AMI_ID),
+		InstanceType: aws.String("t2.micro"),
+		MinCount:     aws.Int64(1),
+		MaxCount:     aws.Int64(1),
+		UserData:     aws.String(base64.StdEncoding.EncodeToString([]byte(userData))),
+	})
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(runResult.Instances) != 1 {
+		t.Fatalf("Expected to launch 1 instance but got %d", len(runResult.Instances))
+	}
+
+	instance := runResult.Instances[0]
+	logger.Printf("Launched instance %s", *instance.InstanceId)
+
+	tagInstance(instance, uniqueId, svc, logger, t)
+
+	return instance, uniqueId
+}
+
+func tagInstance(instance *ec2.Instance, uniqueId string, svc *ec2.EC2, logger *log.Logger, t *testing.T) {
+	logger.Printf("Adding tags to instance %s", *instance.InstanceId)
+
+	_ , err := svc.CreateTags(&ec2.CreateTagsInput{
+		Resources: []*string{instance.InstanceId},
+		Tags: []*ec2.Tag{
+			{
+				Key:   aws.String("Name"),
+				Value: aws.String(fmt.Sprintf("ec2-snapper-unit-test-%s", uniqueId)),
+			},
+		},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func waitForInstanceToStart(instance *ec2.Instance, svc *ec2.EC2, logger *log.Logger, t *testing.T) {
+	logger.Printf("Waiting for instance %s to start...", *instance.InstanceId)
+	if err := svc.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{InstanceIds: []*string{instance.InstanceId}}); err != nil {
+		t.Fatal(err)
+	}
```

Comment:
Heh, the irony is I actually wrote my own version that kept checking the instance status in a loop, with sleep in between... And then while writing some code later on, I stumbled across that function in IntelliJ's auto complete :)